### PR TITLE
feat: structured summary

### PR DESCRIPTION
**Pull request by JonathanStarup**
split summary into logical components for easier maintenance and modification

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/tools/Summary.scala [hunk 1]
```diff
@@ -189,4 +249,31 @@ object Summary {
...
+
+    def addRow(row: List[String]): Unit = {
+      for ((s, i) <- row.iterator.zipWithIndex) {
+        if (i >= maxLens.size) maxLens.append(0)
+        maxLens(i) = maxLens(i) max s.length
+      }
+      rows.append(row)
+    }
+
+    def addRepeatedRow(content: String): Unit = {
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/tools/Summary.scala:266]: Docs for this one would be nice

File: main/src/ca/uwaterloo/flix/tools/Summary.scala [hunk 2]
```diff
@@ -15,168 +15,228 @@
...
+    case object JustIO extends ResEffect
+
+    case object Poly extends ResEffect
+  }
+
+  private sealed trait Function {
+    def loc: SourceLocation
+  }
+
+  private object Function {
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/tools/Summary.scala:216]: Is `Function` the right term for this?

File: main/src/ca/uwaterloo/flix/tools/Summary.scala [hunk 3]
```diff
@@ -15,168 +15,228 @@
...
-    * Returns `true` if the given `spec` is effect polymorphic (neither pure or impure).
-    */
-  private def isEffectPolymorphic(spec: Spec): Boolean = !isPure(spec) && !isUniv(spec)
+  private object FileData {
+    val zero: FileData = FileData(None, 0, 0, 0, 0, 0)
 
-  /**
-    * Right-pads the given string `s` to length `l`.
-    */
+    def combine(l: List[FileData]): FileData = if (l.nonEmpty) l.reduce(_.combine(_)) else zero
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/tools/Summary.scala:181]: Docs for naive sum vs combine

**Review by mlutze — COMMENTED** [review]
Can we also add an example in the docs of what the summary looks like?

I think Magnus would ask for a lot more documentation :)

**Comment by JonathanStarup** [comment]
Good feedback - ive made it better now, more sensible i think

**Review by mlutze — APPROVED** [review]

**Pull request merged**